contracts/libraries/TickMap.sol

```diff
@@ -56,14 +56,16 @@ library TickMap {
 
     function previous(
         IRangePoolStructs.TickMap storage tickMap,
-        int24 tick
+        int24 tick,
+        bool roundUp
     ) internal view returns (
         int24 previousTick
     ) {
         unchecked {
             int24 tickSpacing = IRangePool(address(this)).tickSpacing();
             // rounds up to ensure relative position
-            if (tick % tickSpacing != 0) tick += tickSpacing;
+            if ((tick % tickSpacing != 0 || roundUp)
+                 && tick < _round(TickMath.MAX_TICK, tickSpacing)) tick += tickSpacing;
             (
```

Comment:
Here we round up if `roundUp` is true.

`roundUp` will be true in the case we are searching for the previous tick when swapping.

If `tickAtPrice` is 40 and we need to cross tick 40, we should ask for the previous tick from tick 50 (assume the tickSpacing is 10).